### DISCUSSION TITLE
How to insert compiler attribute `[[intel::reqd_sub_group_size(32)]]`

### DISCUSSION THREAD

**Discussion by alanzhai219** [intel/llvm#13160] [Q&A] [answered]
We all know that sycl kernel can be in functor style or function style or lambda style. 
It is easy to inert `[[intel::reqd_sub_group_size(32)]]` when style is in function style or lambda style.
But how to insert `[[intel::reqd_sub_group_size(32)]]` when launching functor-style kernel.
```cpp
#include <sycl/sycl.hpp>

class vector_add {
public:
    vector_add(float* arg_a, float* arg_b, float* arg_c, size_t arg_N)
        : a(arg_a), b(arg_b), c(arg_c), n(arg_N) {}
    void operator()(sycl::nd_item<3> ndi) const { // attension: must be "const" function
        size_t work_item_idx = ndi.get_local_id(2);
        size_t work_group_idx = ndi.get_group(2);
        size_t work_group_size = ndi.get_local_range(2);
        size_t idx = work_item_idx + work_group_idx * work_group_size;
        if (idx < n) {
            c[idx] = a[idx] + b[idx];
        }
    }
private:
    float* a;
    float* b;
    float* c;
    size_t n;
};

void init_memory(float* ptr, size_t N) {
    for (size_t i=0; i < N; ++i) {
        ptr[i] = i;
    }
}

int main() {
    const size_t N = 256;
    const size_t sub_N = 16;

    // 1. create a queue
    sycl::queue q(sycl::gpu_selector_v);

    // 2. malloc memory
    // 2.1. malloc host memory
    float* a_host = static_cast<float*>(sycl::malloc_host(sizeof(float) * N, q));
    float* b_host = static_cast<float*>(sycl::malloc_host(sizeof(float) * N, q));
    float* c_host = static_cast<float*>(sycl::malloc_host(sizeof(float) * N, q));

    // 2.2. init memory
    init_memory(a_host, N);
    init_memory(b_host, N);
    q.memset(c_host, 0, sizeof(float) * N);

    // 2.3. malloc device memory
    float* a_dev = static_cast<float*>(sycl::malloc_device(sizeof(float) * N, q));
    float* b_dev = static_cast<float*>(sycl::malloc_device(sizeof(float) * N, q));
    float* c_dev = static_cast<float*>(sycl::malloc_device(sizeof(float) * N, q));

    // 3. copy host memory to device memory
    q.memcpy(a_dev, a_host, sizeof(float) * N);
    q.memcpy(b_dev, b_host, sizeof(float) * N);

    // 4. launch kernel
    sycl::range glb_size{1, 1, N};
    sycl::range loc_size{1, 1, sub_N};
    sycl::nd_range parallel_size{glb_size, loc_size};
    // TODO: how to insert [[intel::reqd_sub_group_size(32)]]` here?
    q.parallel_for(parallel_size, [[intel::reqd_sub_group_size(32)]]vector_add(a_dev, b_dev, c_dev, N));
    q.wait();

    // 5. copy device memory to host memory
    q.memcpy(c_host, c_dev, sizeof(float) * N);

    // 6. print
    for (size_t i = 0; i < 10; ++i) {
        std::cout << c_host[i] << "\n";
    }

    // 7. free
    sycl::free(a_host, q);
    sycl::free(b_host, q);
    sycl::free(c_host, q);

    sycl::free(a_dev, q);
    sycl::free(b_dev, q);
    sycl::free(c_dev, q);

    return 0;
}
```

**Comment by AlexeySachkov** [comment, marked as answer]
Hi @alanzhai219,

See [5.8.1. Kernel attributes](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes) for an example how to use attributes with functors

**Reply by alanzhai219** [reply]
Great. Thanks for your reply.
Another question is that this attribute takes effect at compilation right?
Once the attribute is set, the binary or scheduling is fixed.
So, it is should NOT be changed at the runtime stage.
If I am wrong, please correct me. Thanks.

**Reply by AlexeySachkov** [reply]
> Another question is that this attribute takes effect at compilation right?
> Once the attribute is set, the binary or scheduling is fixed.

Right, this "property" can't be later changed at runtime

**Reply by alanzhai219** [reply]
Thanks, it works for me.
BTW, for nvidia gpu device, we all know that warpSize is 32 and developers implement the cuda kernel under warpSize optimization.
But for intel gpu device, it can be configured and software will dispatch the kernel at previous defined grain.
It looks like different value in the property may cause different performance or parallelism.
You know, maybe XE microarch in Arc and PVC is different.
So, is any recommendation for the performance tuning at sub_size level?
Thanks.

**Reply by AlexeySachkov** [reply]
> So, is any recommendation for the performance tuning at sub_size level?

Unfortunately, not from my side, I don't have relevant experience and knowledge.
I will tag @Pennycook here in case he can suggest something, but I guess that there is a lot of trial and error involved.

**Reply by alanzhai219** [reply]
@AlexeySachkov Thanks a lot.

**Reply by Pennycook** [reply]
Right, there isn't a sub-group size that is best for all cases on Intel GPUs.  But I can share my own guidelines.  For these guidelines to make sense, it's important to remember that the sub-group size maps to the SIMD instruction width, and can therefore also be thought of as controlling the number of work-items that are assigned to a single hardware thread.

- If the problem is very small, prefer sub-group size 16.  This doubles the number of hardware threads launched, improving GPU utilization and the ability to hide memory latency.

- If the kernel is very complex, prefer sub-group size 16.  This doubles the number of registers available per work-item (because there are only 16 work-items using the hardware thread's register file instead of 32 work-items), potentially reducing the number of spills.  You can check in the GPU assembly to see if there are spills or not.

- In all other cases, prefer SIMD32.

If you want the flexibility to choose between SIMD16 and SIMD32 for different devices or problem sizes, you can combine the kernel property with a template and runtime dispatch:

```c++
template <typename SubGroupSize>
struct Kernel
{
  // Property accepts template argument as a value because it is a compile-time constant
  void operator()(sycl::nd_item<1> it) [[sycl::reqd_sub_group_size(SubGroupSize)]] {
    // kernel goes here
  }
};

...

if (/* conditions suggest SIMD16 will be best */) {
  q.parallel_for(parallel_size, Kernel<16>());
}
else {
  q.parallel_for(parallel_size, Kernel<32>());
}
```

**Reply by bader** [reply]
FYI: "Mapping Work-Groups to Xe-cores for Maximum Occupancy" chapter of https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2024-0/thread-mapping-and-gpu-occupancy.html

**Reply by alanzhai219** [reply]
@bader Good document.
BTW, the `PDF` layout format is messed up. I didn't find the report issue button. Could you help file a bug for the `PDF` layout format?
Thanks.

**Reply by bader** [reply]
> BTW, the `PDF` layout format is messed up. I didn't find the report issue button. Could you help file a bug for the `PDF` layout format?

@sravanikonda, can you help with that?

**Reply by sravanikonda** [reply]
Thanks for reporting this @alanzhai219. I brought this to the attention of the document owner and they are working on fixing it.

**Reply by alanzhai219** [reply]
@sravanikonda You are welcome. Let's execute with high efficiency.

**Comment by alanzhai219** [comment]
@bader @Pennycook @AlexeySachkov 
Thanks for your comments. That is helpful. I will close this Q&A now. Will re-open if I have more questions about it.
